Social: Move share limits code to Publicize package

danielpost opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
This PR moves the existing share limits code into the Publicize package, and initialises it in Jetpack Social.

Using it in Jetpack will be handled in a follow-up PR.

* Move the share limits code to the Publicize package.
* Introduce a new build step for the Publicize package.

#### Other information:

- [X] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
1202583289493814-as-1202966425123040/f

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

_The testing instructions are very similar to https://github.com/Automattic/jetpack/pull/25661 and https://github.com/Automattic/jetpack/pull/26039_

* Make sure tests pass: `jetpack test plugins/social php` and `jetpack test packages/publicize php`
* Enable the Jetpack Social plugin.
* Add at least two connections to your testing site.
* Make a new post, and check that the sharing screen looks like the first image.
* Enable sharing limits for your site (first line in test plan of D85993-code)

For the block editor:
* Make a new post, and check that you can see the upgrade notice (second image).
* Test that connections are disabled by default when there's more connections than shares left, and that you can't select more connections than allowed.
  * To mock the number of shares left, you can add `$shares['shares_remaining'] = 1;` to `initial_state()` and `$shares_remaining = 1;` to `set_up_sharing_limits()` in `class-publicize-base.php:1502`
* Finally, enable the Jetpack plugin, and make sure the screen looks like the first screenshot again

Then, classic editor:
* Deactivate Jetpack again.
* Activate the Classic Editor plugin.
* Start writing a new post.
* Make sure the notice shows with the number of shares left and the upgrade link.
* To mock the number of shares left, you can add `$shares['shares_remaining'] = 1;` to `initial_state()` and `$shares_remaining = 1;` to `set_up_sharing_limits()` in `class-publicize-base.php:1502`
* Start a new post again, and check the following:
    * All connections are disabled.
    * You can only check one connection at a time.
<table>
<tr>
	<td>Sharing Limits Disabled 
	<td>Sharing Limits Enabled
	<td>More connections than shares left
	<td>More connections than shares left, and all remaining shares have been used
<tr>
	<td><img width="278" alt="image" src="https://user-images.githubusercontent.com/1713699/184976111-03be1d83-04b1-449c-bdea-583b8281a292.png">
	<td><img width="278" alt="image" src="https://user-images.githubusercontent.com/1713699/184976450-e0c32894-9eb3-4402-a5ea-826f367c2d25.png">
	<td><img width="280" alt="image" src="https://user-images.githubusercontent.com/1713699/184976463-949ade72-0653-4c15-95a3-e20fa62945a4.png">
	<td><img width="280" alt="image" src="https://user-images.githubusercontent.com/1713699/184976475-d15d5279-9561-4d0f-b536-f7b34e431524.png">
</table>

Commit:
[not verified] Move Classic Editor script to Publicize package

Commit:
[not verified] Move share limits script to Publicize package

Commit:
[not verified] # This is a combination of 2 commits.

[not verified] Move init to Publicize class

changelog

Commit:
[not verified] Move tests

pablinos commented on projects/plugins/social/src/class-jetpack-social.php:
```diff
@@ -299,21 +299,9 @@ public function set_up_sharing_limits() {
-		$shares_remaining = $info['shares_remaining'];
-		$share_limits     = new Automattic\Jetpack\Social\Share_Limits( $connections, $shares_remaining );
-		$share_limits->enforce_share_limits();
+		$publicize->set_up_sharing_limits();
```
Should we move the `has_paid_plan` check into this method? We could also add something to make sure it's only run on certain admin requests, although that might need a filter, so we can add the Social admin screen to it.

pablinos replied:
Thinking about this some more I wonder if this logic should be in [the setup code that's run by the Config package](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/publicize/src/class-publicize-setup.php). It's slightly tricky as that isn't run on WPCOM currently, but that does seem like the natural place for setup code, and it would handle both plugins at the same time.

pablinos commented on projects/packages/publicize/webpack.config.js:
```diff
@@ -0,0 +1,55 @@
+const path = require( 'path' );
```
It's a shame that we have to add all of this config for just for the classic editor script, but I suppose it can't be helped!

pablinos commented on pnpm-lock.yaml:
```diff
@@ -7131,7 +7140,6 @@ packages:
 
   /@types/estree/1.0.0:
     resolution: {integrity: sha512-WulqXMDUTYAXCjZnk6JtIHPigp55cVtDgDrO2gHRwhyJto21+1zbVCtOYB2L1F9w4qCQ0rOGWBnBe0FNTiEJIQ==}
-    dev: true
```
Do you know why this is removed? I can't see what we're doing that would make it not a dev dependency.

pablinos reviewed (commented):
This is working well for me. Just a couple of thoughts really.

We can sort out adding it Jetpack in a follow-up

Commit:
[not verified] Move initialising code to setup class

spsiddarthan committed:
[not verified] fix share limits init code.

spsiddarthan committed:
[not verified] passs screen as a param to sharing limits class.

pablinos committed:
[not verified] Swap to passing the current editor.

pablinos commented on projects/packages/publicize/webpack.config.js:
```diff
@@ -0,0 +1,55 @@
+				} ),
+
+				// Handle CSS.
+				jetpackWebpackConfig.CssRule( {
+					extensions: [ 'css', 'sass', 'scss' ],
+					extraLoaders: [
+						{
+							loader: 'postcss-loader',
+							options: {
+								postcssOptions: { config: path.join( __dirname, 'postcss.config.js' ) },
+							},
+						},
+						'sass-loader',
+					],
+				} ),
```
I wonder if this is necessary? I don't think we have a postcss.config.js file here. We might be able to simplify this config a bit more.

pablinos replied:
We're going to look at this in a follow up.

pablinos committed:
[not verified] # This is a combination of 6 commits.

[not verified] Update project dependencies.

changelog

Update package dependency

Fix the tests.

Update the pnpm-lock file.

More dependency hell.

pablinos reviewed (commented):
I think this is pretty close now. I just wondered if we could simplify the webpack config.

pablinos commented on projects/packages/publicize/src/class-publicize-base.php:
```diff
@@ -1486,6 +1487,19 @@ public function calculate_scheduled_shares( $blog_id ) {
+	public function has_paid_plan() {
+		static $has_paid_plan = null;
+		if ( ! $has_paid_plan ) {
+			$has_paid_plan = Current_Plan::supports( 'social-shares-1000', true );
```
We should remove that `true` parameter and work out tapping into the mechanism that calls into the site to update the plan info. We'll be creating a lot of extra API requests here for something that hardly ever changes.

spsiddarthan replied:
That makes sense, @pablinos! We need to be wary of a few things though - if someone's paid plan expires, and is over the limit, they shouldn't be able to share. 

Additionally, if someone is on a free plan and has hit the limit when they're redirected to the editor after buying the plan, they should not see the nudge.

pablinos replied:
Yes, good point. I'm not sure what we do as the plan expires, but it must be a similar thing. I think being redirected back to the editor would be fine, depending on the sequence of events. If they're redirected after the callback is made, then everything should be fine, otherwise there might be some kind of race condition. 

Do you think we could add a parameter to the redirect that would allow the plan information to be updated?

For a starter, it would be good if we only pass `true` here for Jetpack Social, and work out caching the information on Jetpack and allowing WPCOM to passthrough.

pablinos commented on projects/packages/publicize/src/class-publicize-setup.php:
```diff
@@ -22,18 +22,52 @@ public static function configure() {
+			return;
+		}
+
+		$info = $publicize->get_publicize_shares_info( \Jetpack_Options::get_option( 'id' ) );
```
Two more ponderings here.

- It would be good to re-use this init function on WPCOM, and so we'd need to use `get_current_blog_id()` here. As we probably wouldn't be calling it on the `current_screen` hook, we could maybe pass the blog ID in, and if it's null/0 default to the Jetpack_Options?
- `get_publicize_shares_info` will make an API call everytime. As we roll this out, is there a way to cache the feature flag value locally (when it's false and when we don't receive the shares info, as that shows we're not interested in it)? If we store it in a transient for a long time, we could invalidate it on [the `jetpack_get_site_data_success` hook](
https://github.com/Automattic/jetpack/tree/d5dcf6fab60220ed844d52f162c6bd9c7bac5bca/projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php#L1935), so it gets refreshed each time the next time after the site data has been refreshed.

pablinos reviewed (commented):
I had a couple more thoughts about this. They're partly 'notes to self' as I was looking at implementing this in Jetpack

pablinos committed:
[not verified] Futureproof init_sharing_limits for use on WPCOM.

pablinos committed:
Update composer.lock files.

pablinos committed:
Cache the result of the API shares info call.

pablinos committed:
Fix the caching of the shares info.

spsiddarthan committed:
[not verified] Don't force refresh plan data for Jetpack.

spsiddarthan committed:
Force refresh plan data for Jetpack Social.

spsiddarthan committed:
Remove check for Jetpack class

spsiddarthan committed:
Fix response sent back by shares response.

spsiddarthan committed:
Remove the else block for shares-info.

pablinos committed:
Make the plan refresh configurable as the package is set up.

pablinos committed:
Reinstate the wp_error check.

pablinos committed:
Merge remote-tracking branch 'origin/trunk' into change/move-sharing-limits-to-publicize-package

pablinos committed:
Merge branch 'trunk' into change/move-sharing-limits-to-publicize-package

danielpost commented:
FYI: I can't approve this since I'm the original PR author, but I've just gone through all the test steps again and reviewed the latest code changes—I think it's good to merge!

pablinos committed:
Merge branch 'trunk' into change/move-sharing-limits-to-publicize-package

pablinos reviewed (dismissed):
Let's get this merged, and we can carry on in the follow-up #26420

pablinos committed:
Merge remote-tracking branch 'origin/trunk' into change/move-sharing-limits-to-publicize-package

pablinos committed:
Update the package versions/dependencies.

pablinos committed:
Merge branch 'trunk' into change/move-sharing-limits-to-publicize-package

pablinos committed:
Merge branch 'trunk' into change/move-sharing-limits-to-publicize-package

pablinos reviewed (approved):
I'm re-approving this, but leaving it with 'Needs Review' for now in case someone has a chance to cast an extra eye over it before the morning.

kraftbj commented:
r253241-wpcom
